File: backend/app/services/delivery/google_calendar.py
"""Google Calendar appointment sync with prepare / call / finalize.

**Idempotent recovery:** create uses a deterministic Calendar event ``id`` derived
from the appointment id (charset ``[a-v0-9]{5,1024}``). On HTTP 409 conflict,
GET that event and attach. Secondary recovery searches private extended
property ``immigromeAppointmentId``. Timeouts after create may leave an event
that reclaim reattaches without creating a duplicate.

Unlike Telegram (at-most-once automatic delivery), Calendar may safely retry
the external call because provider-side identity is deterministic.
"""

from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Any, Literal
from urllib.parse import quote

import httpx
import jwt
from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import get_settings
from app.db.tables import appointment_table

CALENDAR_SCOPE = "https://www.googleapis.com/auth/calendar.events"
_GOOGLE_EVENT_ID_RE = re.compile(r"^[a-v0-9]{5,1024}$")


def deterministic_google_event_id(appointment_id: str) -> str:
    """Map appointment id → Google-allowed client-supplied event id."""
    digest = hashlib.sha256(
        f"immigrome-appointment:{appointment_id}".encode()
    ).hexdigest()
    # hexdigest is [0-9a-f], a subset of [a-v0-9], length 64.
    if not _GOOGLE_EVENT_ID_RE.match(digest):
        raise RuntimeError("deterministic Google event id failed charset check")
    return digest


@dataclass(frozen=True)
class PreparedCalendarCreate:
    action: Literal["create", "skip"]
    appointment_id: str
    title: str | None
    starts_at: datetime | None
    ends_at: datetime | None
    timezone: str | None
    attendees: tuple[dict[str, str], ...]
    calendar_id: str | None
    credentials_json: str | None
    provider_event_id: str | None


@dataclass(frozen=True)
class CalendarCallResult:
    google_event_id: str


async def _access_token(credentials: dict[str, Any]) -> str:
    now = datetime.now(UTC)
    token_uri = str(credentials.get("token_uri") or "https://oauth2.googleapis.com/token")
    assertion = jwt.encode(
        {
            "iss": credentials["client_email"],
            "scope": CALENDAR_SCOPE,
            "aud": token_uri,
            "iat": int(now.timestamp()),
            "exp": int((now + timedelta(minutes=55)).timestamp()),
        },
        credentials["private_key"],
        algorithm="RS256",
    )
    async with httpx.AsyncClient(timeout=20.0) as client:
        response = await client.post(
            token_uri,
            data={
                "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
                "assertion": assertion,
            },
        )
    response.raise_for_status()
    token = response.json().get("access_token")
    if not token:
        raise RuntimeError("Google OAuth did not return an access token")
    return str(token)


async def find_google_event_by_appointment(
    client: httpx.AsyncClient,
    *,
    token: str,
    calendar_id: str,
    appointment_id: str,
) -> str | None:
    url = (
        "https://www.googleapis.com/calendar/v3/calendars/"
        f"{quote(calendar_id, safe='')}/events"
    )
    response = await client.get(
        url,
        headers={"Authorization": f"Bearer {token}"},
        params={
            "privateExtendedProperty": f"immigromeAppointmentId={appointment_id}",
            "maxResults": 1,
            "singleEvents": "true",
        },
    )
    response.raise_for_status()
    items = response.json().get("items")
    if not isinstance(items, list) or not items:
        return None
    event_id = items[0].get("id") if isinstance(items[0], dict) else None
    return str(event_id) if event_id else None


async def get_google_event_by_id(
    client: httpx.AsyncClient,
    *,
    token: str,
    calendar_id: str,
    event_id: str,
) -> str | None:
    url = (
        "https://www.googleapis.com/calendar/v3/calendars/"
        f"{quote(calendar_id, safe='')}/events/{quote(event_id, safe='')}"
    )
    response = await client.get(
        url,
        headers={"Authorization": f"Bearer {token}"},
    )
    if response.status_code == 404:
        return None
    response.raise_for_status()
    payload = response.json()
    found = payload.get("id")
    return str(found) if found else None


async def prepare_google_calendar_event(
    db: AsyncSession, event: dict[str, Any]
) -> PreparedCalendarCreate:
    """Load appointment; skip if already synced. No HTTP."""
    appointment_id = str(event["payloadJson"].get("appointment_id") or "")
    appointment = (
        await db.execute(
            select(appointment_table).where(appointment_table.c.id == appointment_id)
        )
    ).mappings().first()
    if appointment is None:
        raise ValueError("Appointment not found")
    if appointment.googleEventId:
        return PreparedCalendarCreate(
            action="skip",
            appointment_id=appointment_id,
            title=None,
            starts_at=None,
            ends_at=None,
            timezone=None,
            attendees=(),
            calendar_id=None,
            credentials_json=None,
            provider_event_id=None,
        )
    settings = get_settings()
    if not settings.google_calendar_id or not settings.google_service_account_json:
        raise RuntimeError("Google Calendar is not configured")
    participants = appointment.participantsJson or []
    attendees = tuple(
        {"email": str(item["email"])}
        for item in participants
        if isinstance(item, dict) and item.get("email")
    )
    return PreparedCalendarCreate(
        action="create",
        appointment_id=appointment_id,
        title=appointment.title,
        starts_at=appointment.startsAt,
        ends_at=appointment.endsAt,
        timezone=appointment.timezone,
        attendees=attendees,
        calendar_id=settings.google_calendar_id,
        credentials_json=settings.google_service_account_json,
        provider_event_id=deterministic_google_event_id(appointment_id),
    )


async def call_google_calendar_event(
    prepared: PreparedCalendarCreate,
) -> CalendarCallResult:
    """OAuth + Calendar HTTP outside any DB transaction."""
    if prepared.action != "create":
        raise RuntimeError("call_google_calendar_event requires action=create")
    assert (
        prepared.calendar_id
        and prepared.credentials_json
        and prepared.provider_event_id
        and prepared.title
        and prepared.starts_at
        and prepared.ends_at
        and prepared.timezone
    )
    credentials = json.loads(prepared.credentials_json)
    token = await _access_token(credentials)
    calendar_id = prepared.calendar_id
    provider_event_id = prepared.provider_event_id
    url = (
        "https://www.googleapis.com/calendar/v3/calendars/"
        f"{quote(calendar_id, safe='')}/events"
    )
    attendees = list(prepared.attendees)
    async with httpx.AsyncClient(timeout=20.0) as client:
        # Primary: create with deterministic id.
        try:
            response = await client.post(
                url,
                headers={"Authorization": f"Bearer {token}"},
                params={"sendUpdates": "all" if attendees else "none"},
                json={
                    "id": provider_event_id,
                    "summary": prepared.title,
                    "start": {
                        "dateTime": prepared.starts_at.isoformat(),
                        "timeZone": prepared.timezone,
                    },
                    "end": {
                        "dateTime": prepared.ends_at.isoformat(),
                        "timeZone": prepared.timezone,
                    },
                    "attendees": attendees,
                    "extendedProperties": {
                        "private": {
                            "immigromeAppointmentId": prepared.appointment_id
                        }
                    },
                },
            )
        except (httpx.TimeoutException, httpx.NetworkError):
            recovered = await get_google_event_by_id(
                client,
                token=token,
                calendar_id=calendar_id,
                event_id=provider_event_id,
            )
            if recovered:
                return CalendarCallResult(google_event_id=recovered)
            secondary = await find_google_event_by_appointment(
                client,
                token=token,
                calendar_id=calendar_id,
                appointment_id=prepared.appointment_id,
            )
            if secondary:
                return CalendarCallResult(google_event_id=secondary)
            raise

        if response.status_code == 409:
            recovered = await get_google_event_by_id(
                client,
                token=token,
                calendar_id=calendar_id,
                event_id=provider_event_id,
            )
            if recovered:
                return CalendarCallResult(google_event_id=recovered)
            secondary = await find_google_event_by_appointment(
                client,
                token=token,
                calendar_id=calendar_id,
                appointment_id=prepared.appointment_id,
            )
            if secondary:
                return CalendarCallResult(google_event_id=secondary)
            response.raise_for_status()

        response.raise_for_status()
        google_event_id = response.json().get("id")
        if not google_event_id:
            raise RuntimeError("Google Calendar did not return an event id")
        return CalendarCallResult(google_event_id=str(google_event_id))


async def finalize_google_calendar_event(
    db: AsyncSession,
    prepared: PreparedCalendarCreate,
    *,
    result: CalendarCallResult | None = None,
) -> None:
    """Persist googleEventId after a successful call. No HTTP."""
    if prepared.action == "skip":
        return
    if result is None:
        raise RuntimeError("finalize_google_calendar_event requires result")
    now = datetime.now(UTC)
    await db.execute(
        update(appointment_table)
        .where(appointment_table.c.id == prepared.appointment_id)
        .values(
            googleEventId=result.google_event_id,
            status="CONFIRMED",
            updatedAt=now,
        )
    )


async def create_google_calendar_event(
    db: AsyncSession, event: dict[str, Any]
) -> None:
    """Compatibility helper; prefer prepare → call → finalize in the worker."""
    prepared = await prepare_google_calendar_event(db, event)
    if prepared.action == "skip":
        return
    await db.flush()
    result = await call_google_calendar_event(prepared)
    await finalize_google_calendar_event(db, prepared, result=result)

File: src/server/delivery/google-calendar.ts
import { createHash, createSign, randomUUID } from "crypto";
import type { Appointment, OutboxEvent } from "@prisma/client";
import { prisma } from "@/lib/db";

const CALENDAR_SCOPE = "https://www.googleapis.com/auth/calendar.events";
const GOOGLE_EVENT_ID_RE = /^[a-v0-9]{5,1024}$/;

export type ServiceAccountCredentials = {
  client_email: string;
  private_key: string;
  token_uri?: string;
};

export type PreparedCalendarUpsert = {
  action: "create" | "patch" | "skip";
  appointment: Appointment;
  providerEventId: string;
  calendarId: string;
  credentials: ServiceAccountCredentials;
};

export type PreparedCalendarDelete = {
  action: "delete" | "skip";
  appointmentId: string;
  googleEventId: string | null;
  calendarId: string | null;
  credentials: ServiceAccountCredentials | null;
};

export function deterministicGoogleEventId(appointmentId: string): string {
  const digest = createHash("sha256")
    .update(`immigrome-appointment:${appointmentId}`)
    .digest("hex");
  if (!GOOGLE_EVENT_ID_RE.test(digest)) {
    throw new Error("deterministic Google event id failed charset check");
  }
  return digest;
}

export function parseServiceAccountJson(
  raw: string | undefined,
): ServiceAccountCredentials {
  if (!raw?.trim()) {
    throw new Error("GOOGLE_SERVICE_ACCOUNT_JSON is not configured");
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    throw new Error("GOOGLE_SERVICE_ACCOUNT_JSON is not valid JSON");
  }
  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new Error("GOOGLE_SERVICE_ACCOUNT_JSON must be an object");
  }
  const obj = parsed as Record<string, unknown>;
  const clientEmail = obj.client_email;
  const privateKey = obj.private_key;
  if (typeof clientEmail !== "string" || typeof privateKey !== "string") {
    throw new Error(
      "GOOGLE_SERVICE_ACCOUNT_JSON requires client_email and private_key",
    );
  }
  return {
    client_email: clientEmail,
    private_key: privateKey.replace(/\\n/g, "\n"),
    token_uri:
      typeof obj.token_uri === "string"
        ? obj.token_uri
        : "https://oauth2.googleapis.com/token",
  };
}

function base64url(input: Buffer | string): string {
  const buf = typeof input === "string" ? Buffer.from(input) : input;
  return buf
    .toString("base64")
    .replace(/=/g, "")
    .replace(/\+/g, "-")
    .replace(/\//g, "_");
}

export async function getGoogleAccessToken(
  credentials: ServiceAccountCredentials,
  fetchImpl: typeof fetch = fetch,
): Promise<string> {
  const now = Math.floor(Date.now() / 1000);
  const tokenUri =
    credentials.token_uri ?? "https://oauth2.googleapis.com/token";
  const header = base64url(JSON.stringify({ alg: "RS256", typ: "JWT" }));
  const claim = base64url(
    JSON.stringify({
      iss: credentials.client_email,
      scope: CALENDAR_SCOPE,
      aud: tokenUri,
      iat: now,
      exp: now + 55 * 60,
    }),
  );
  const unsigned = `${header}.${claim}`;
  const signer = createSign("RSA-SHA256");
  signer.update(unsigned);
  signer.end();
  const signature = base64url(signer.sign(credentials.private_key));
  const assertion = `${unsigned}.${signature}`;

  const response = await fetchImpl(tokenUri, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "urn:ietf:params:oauth:grant-type:jwt-bearer",
      assertion,
    }),
  });
  const data = (await response.json().catch(() => null)) as {
    access_token?: string;
    error?: string;
  } | null;
  if (!response.ok || !data?.access_token) {
    throw new Error(
      data?.error ?? `Google OAuth token exchange failed (${response.status})`,
    );
  }
  return data.access_token;
}

function calendarEventsUrl(calendarId: string, eventId?: string): string {
  const base = `https://www.googleapis.com/calendar/v3/calendars/${encodeURIComponent(calendarId)}/events`;
  return eventId ? `${base}/${encodeURIComponent(eventId)}` : base;
}

function eventBody(appointment: Appointment, providerEventId: string) {
  return {
    id: providerEventId,
    summary: appointment.title,
    start: {
      dateTime: appointment.startsAt.toISOString(),
      timeZone: appointment.timezone,
    },
    end: {
      dateTime: appointment.endsAt.toISOString(),
      timeZone: appointment.timezone,
    },
    extendedProperties: {
      private: {
        immigromeAppointmentId: appointment.id,
      },
    },
  };
}

export async function prepareCalendarUpsert(
  event: OutboxEvent,
  env: NodeJS.ProcessEnv | Record<string, string | undefined> = process.env,
): Promise<PreparedCalendarUpsert> {
  const payload = event.payloadJson;
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    throw new Error("calendar.upsert payload must be an object");
  }
  const appointmentId = String(
    (payload as { appointmentId?: unknown }).appointmentId ?? "",
  );
  if (!appointmentId) {
    throw new Error("calendar.upsert requires appointmentId");
  }

  const appointment = await prisma.appointment.findUnique({
    where: { id: appointmentId },
  });
  if (!appointment) {
    throw new Error(`Appointment ${appointmentId} not found`);
  }
  if (appointment.status === "CANCELLED") {
    throw new Error("Cannot upsert a cancelled appointment");
  }

  const calendarId = env.GOOGLE_CALENDAR_ID?.trim();
  if (!calendarId) {
    throw new Error("GOOGLE_CALENDAR_ID is not configured");
  }
  const credentials = parseServiceAccountJson(env.GOOGLE_SERVICE_ACCOUNT_JSON);
  const providerEventId =
    appointment.googleEventId ?? deterministicGoogleEventId(appointment.id);

  if (
    appointment.googleEventId &&
    appointment.status === "CONFIRMED" &&
    appointment.googleEventId === providerEventId
  ) {
    // Still patch on reschedule (PENDING after reschedule). Skip only if already confirmed
    // and outbox is a pure duplicate with matching version semantics handled by idempotency key.
  }

  const action: PreparedCalendarUpsert["action"] = appointment.googleEventId
    ? "patch"
    : "create";

  return {
    action,
    appointment,
    providerEventId,
    calendarId,
    credentials,
  };
}

export async function callCalendarUpsert(
  prepared: PreparedCalendarUpsert,
  fetchImpl: typeof fetch = fetch,
): Promise<{ googleEventId: string }> {
  const token = await getGoogleAccessToken(prepared.credentials, fetchImpl);
  const body = eventBody(prepared.appointment, prepared.providerEventId);

  if (prepared.action === "patch" && prepared.appointment.googleEventId) {
    const response = await fetchImpl(
      calendarEventsUrl(prepared.calendarId, prepared.appointment.googleEventId),
      {
        method: "PATCH",
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          summary: body.summary,
          start: body.start,
          end: body.end,
          extendedProperties: body.extendedProperties,
        }),
      },
    );
    if (response.ok) {
      const data = (await response.json()) as { id?: string };
      return { googleEventId: String(data.id ?? prepared.appointment.googleEventId) };
    }
    if (response.status !== 404) {
      throw new Error(`Google Calendar PATCH failed (${response.status})`);
    }
    // Fall through to create if patch target missing.
  }

  const insertResponse = await fetchImpl(calendarEventsUrl(prepared.calendarId), {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  if (insertResponse.status === 409) {
    const getResponse = await fetchImpl(
      calendarEventsUrl(prepared.calendarId, prepared.providerEventId),
      {
        headers: { Authorization: `Bearer ${token}` },
      },
    );
    if (getResponse.ok) {
      const data = (await getResponse.json()) as { id?: string };
      if (data.id) return { googleEventId: String(data.id) };
    }

    const search = new URL(calendarEventsUrl(prepared.calendarId));
    search.searchParams.set(
      "privateExtendedProperty",
      `immigromeAppointmentId=${prepared.appointment.id}`,
    );
    search.searchParams.set("maxResults", "1");
    search.searchParams.set("singleEvents", "true");
    const searchResponse = await fetchImpl(search.toString(), {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (searchResponse.ok) {
      const data = (await searchResponse.json()) as {
        items?: Array<{ id?: string }>;
      };
      const found = data.items?.[0]?.id;
      if (found) return { googleEventId: String(found) };
    }
    throw new Error("Google Calendar 409 conflict and recovery failed");
  }

  if (!insertResponse.ok) {
    throw new Error(`Google Calendar INSERT failed (${insertResponse.status})`);
  }
  const data = (await insertResponse.json()) as { id?: string };
  if (!data.id) {
    throw new Error("Google Calendar INSERT returned no event id");
  }
  return { googleEventId: String(data.id) };
}

export async function finalizeCalendarUpsert(
  appointmentId: string,
  googleEventId: string,
): Promise<Appointment> {
  // Client already confirmed via Telegram/admin; calendar sync marks CONFIRMED.
  return prisma.appointment.update({
    where: { id: appointmentId },
    data: {
      googleEventId,
      status: "CONFIRMED",
    },
  });
}

export async function prepareCalendarDelete(
  event: OutboxEvent,
  env: NodeJS.ProcessEnv | Record<string, string | undefined> = process.env,
): Promise<PreparedCalendarDelete> {
  const payload = event.payloadJson;
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    throw new Error("calendar.delete payload must be an object");
  }
  const appointmentId = String(
    (payload as { appointmentId?: unknown }).appointmentId ?? "",
  );
  if (!appointmentId) {
    throw new Error("calendar.delete requires appointmentId");
  }

  const appointment = await prisma.appointment.findUnique({
    where: { id: appointmentId },
  });
  if (!appointment) {
    return {
      action: "skip",
      appointmentId,
      googleEventId: null,
      calendarId: null,
      credentials: null,
    };
  }

  const googleEventId =
    appointment.googleEventId ??
    (typeof (payload as { googleEventId?: unknown }).googleEventId === "string"
      ? String((payload as { googleEventId?: unknown }).googleEventId)
      : null);

  if (!googleEventId) {
    return {
      action: "skip",
      appointmentId,
      googleEventId: null,
      calendarId: null,
      credentials: null,
    };
  }

  const calendarId = env.GOOGLE_CALENDAR_ID?.trim() ?? null;
  if (!calendarId) {
    throw new Error("GOOGLE_CALENDAR_ID is not configured");
  }
  const credentials = parseServiceAccountJson(env.GOOGLE_SERVICE_ACCOUNT_JSON);

  return {
    action: "delete",
    appointmentId,
    googleEventId,
    calendarId,
    credentials,
  };
}

export async function callCalendarDelete(
  prepared: PreparedCalendarDelete,
  fetchImpl: typeof fetch = fetch,
): Promise<void> {
  if (
    prepared.action !== "delete" ||
    !prepared.googleEventId ||
    !prepared.calendarId ||
    !prepared.credentials
  ) {
    return;
  }
  const token = await getGoogleAccessToken(prepared.credentials, fetchImpl);
  const response = await fetchImpl(
    calendarEventsUrl(prepared.calendarId, prepared.googleEventId),
    {
      method: "DELETE",
      headers: { Authorization: `Bearer ${token}` },
    },
  );
  if (!response.ok && response.status !== 404 && response.status !== 410) {
    throw new Error(`Google Calendar DELETE failed (${response.status})`);
  }
}

export async function finalizeCalendarDelete(
  appointmentId: string,
): Promise<void> {
  await prisma.appointment.updateMany({
    where: { id: appointmentId },
    data: {
      googleEventId: null,
      status: "CANCELLED",
    },
  });
}

/** Test helper: unique client request id. */
export function newAppointmentClientRequestId(): string {
  return `appt:${randomUUID()}`;
}
